### ISSUE TITLE
MAC address as part of hunter send data itself?

### ISSUE THREAD

**Issue by ColonelPanicks** [openflighthpc/flight-hunter#45]

**Comment by Womblue** [comment]
It seems that the MAC address can't be collected from the client in the same way that we can get its IP. As far as I can see there are two main ways we could go about giving the MAC data to hunter:

- Grab it manually by parsing the `ip neigh` command's output.
- Obtain it from Gather's output

If we decide to use the output from Gather it would involve some further integration of Hunter and Gather. In the same manner that Profile can optionally use Hunter, I think the most logical choice is to give Hunter an option to use Gather and integrate the Gather data into Hunter's sent JSON instead of simply sending the YAML string as the payload. This would give Hunter direct access to the gather data, allowing the MAC addresses (and any future required info) to be associated with Hunter nodes, and would also make the sending of gather data via the `send` command considerably simpler.

**Comment by VoxSecundus** [comment]
@ColonelPanicks When you refer to a system's MAC address, I assume you're looking for the MAC address of the interface holding the IP that is able to resolve the `target_host` host. 

@Womblue is right in that obtaining network interface info introduces further coupling with Flight Gather, which is something I'd like to avoid at the moment. I'm not against coupling them more comprehensively in the future, but to have Hunter depend on Gather for something that it can likely do alone is pointless.

Socket APIs at the transport/network layers have no understanding of the data link later, so the server cannot directly infer which network interface sent the packet.

I believe that the solution lies with the client sending some [truncated info about its own network interfaces](https://stackoverflow.com/a/39111852). By doing so, the server machine can cross reference the client's `peeraddr` with the interface data sent by the client. I realise that this looks like a duplication of interface data being sent by Gather, but we must remember that Gather technically has nothing to do with Hunter; we are just using it to generate a block of text that Hunter sends. Any data duplication that occurs is the burden of the user.

**Comment by ColonelPanicks** [comment]
You are correct that I am looking for the MAC address of the interface matching the sender's IP. I had (maybe naively) hoped that this would be something we could infer from the packet itself. 

On the note of gather information I think too much tight coupling wouldn't be a wise move right now and, in terms of sending further info, I think that's where gather and amorphous content works well. The MAC address fits into the idea of some basic sender info that assists with identifying the clients. A further benefit of MAC being shown is when, for example, manually hunting multiple physical servers - in our internal use case we can easily cross reference the MAC address printed on the physical machines ID tag to the hunter logs to identify which one it is. 

Another thought is that I think sending all MAC addresses for the sender would be overkill and would further muddy some waters.

**Comment by ColonelPanicks** [comment]
My only thinking against using it would be it involves shelling out from hunter but that's not the end of the world (and perhaps there are some cleaner ways of accessing details in the same way). 

My other thought is wondering what the restrictions may be (e.g. from firewall/network security perspective) on this returning info. That concern would be put to rest by a test on AWS/Azure to ensure their network doesn't prevent this sort of stuff. 

Basically, `ip neigh/arp` seems like a decent solution to this problem!
